lib/IDE/CodeCompletion.cpp

```diff
@@ -976,9 +981,51 @@ void CodeCompletionResultBuilder::addCallParameter(Identifier Name,
         PrintOptions::OpaqueReturnTypePrintingMode::WithoutOpaqueKeyword;
     if (ContextTy)
       PO.setBaseType(ContextTy);
-    addChunkWithText(
-        CodeCompletionString::Chunk::ChunkKind::CallParameterClosureType,
-        AFT->getString(PO));
+
+    if (isLabeledTrailingClosure) {
+      // Expand the closure body.
+      SmallString<32> buffer;
+      llvm::raw_svector_ostream OS(buffer);
+
+      bool returnsVoid = AFT->getResult()->isVoid();
+      bool hasSignature = !returnsVoid || !AFT->getParams().empty();
+      if (hasSignature)
+        OS << "(";
+      bool firstParam = true;
+      for (const auto &param : AFT->getParams()) {
+        if (!firstParam)
+          OS << ", ";
+        firstParam = false;
+
+        if (param.hasLabel()) {
+          OS << param.getLabel();
+        } else {
+          OS << "<#";
+          if (param.isInOut())
+            OS << "inout ";
+          OS << param.getPlainType()->getString(PO);
+          if (param.isVariadic())
+            OS << "...";
+          OS << "#>";
+        }
+      }
+      if (hasSignature)
+        OS << ")";
+      if (!returnsVoid)
+        OS << " -> " << AFT->getResult()->getString(PO);
```

Comment:
@benlangmuir If the return type is unresolved type, could you stop emitting `-> _`?

Reply:
Do you have a suggestion for a test case? Everything I've tried either gives a name or it fails entirely to provide argument completions.

Reply:
Maybe a generic param type or a archetype type. Sorry I don't have this branch in my hand, but does this suggests `y: { () -> A in <#code#> }`? 
```swift
func foo<A>(x: () -> Void, y: () -> A) {}
foo() {} #^COMPLETE^#
```
If so, I think we should not suggest the signature in this case.

Reply:
Ah yep it gives `y: { () -> A in <#code#> }`.  This is consistent with how completing the whole "foo" call behaves though (if you complete foo and then expand the placeholder the result is identical), so if we're going to change it I think it should be a separate PR.

In general, we might want to drop the return type in most cases even if it is resolved - it seems to be more common style to omit the types unless they are required.  We could also drop the parens around the parameters in that case.

Reply:
> I think it should be a separate PR

Agreed!